src/audio/component.c

```diff
@@ -29,31 +29,68 @@ DECLARE_SOF_UUID("component", comp_uuid, 0x7c42ce8b, 0x0108, 0x43d0,
 
 DECLARE_TR_CTX(comp_tr, SOF_UUID(comp_uuid), LOG_LEVEL_INFO);
 
-static const struct comp_driver *get_drv(uint32_t type)
+static const struct comp_driver *get_drv(struct sof_ipc_comp *comp)
 {
 	struct comp_driver_list *drivers = comp_drivers_get();
 	struct list_item *clist;
 	const struct comp_driver *drv = NULL;
 	struct comp_driver_info *info;
-	uint32_t flags;
+	struct sof_ipc_comp_new_ext *comp_ext;
 
-	irq_local_disable(flags);
+	if (comp->ext_data_offset &&
+	    comp->hdr.size <
+	    comp->ext_data_offset + sizeof(struct sof_ipc_comp_new_ext)) {
+		tr_err(&comp_tr, "get_drv(): header mismatched! ext_data_offset:0x%x, hdr.size: 0x%x",
+		       comp->ext_data_offset, comp->hdr.size);
+		return NULL;
+	}
+
+	/* use UUID to match the driver if UUID is provided */
+	if (comp->ext_data_offset) {
+		comp_ext = (struct sof_ipc_comp_new_ext *)
+			   ((uint8_t *)comp + comp->ext_data_offset);
+
+		/* use component type if old tplg without UUID used */
+		if (sof_is_uuid_nil(comp_ext->uuid)) {
```

Comment:
Dont need this, If extended data exists then UUID is considered valid.

Reply:
Not actually, the extended data will be nil if it is not provided with topology UUID token at the moment. And furthermore, if more extended data is added, there could be no UUID but with other data only.

Reply:
I think we should say extended data is ABI3.17, so UUID is compulsary after that point. i.e. if extended data exists then UUID exists.

Reply:
I meant the UUID array is included in the extended data struct, but the UUID token (e.g. for some specific component, or old tplg file is used), in this scenario, the driver will fill Nil (all 0s) to the extended data UUID array, to keep it backward compatible, we need to filter out this case and fall back to use component type for it.

Reply:
extended data and hence UUID is only sent to FW iff topology AND kernel are ABI >= 3.17. Kernel will not send extended data for topology prior to ABI 3.17. Please fix and resend.
@kv2019i how is the kernel PR progressing for UUID ? The FW part is almost ready to merge.

Reply:
@lgirdwood it's moving, but a few points of debate left. If nothing new comes up, we should be ready end-of-week.

Reply:
someone will have to convince me why we are dealing with a flexible array without using a flexible array definition.

Reply:
@plbossart sorry, what flex array are you referring too ? I cant see any on this PR unless the kernel PR is using one ?

Reply:
@plbossart (and @marc-hb @lgirdwood ) it's not a flex array. We basicly have a extension problem here which we have to workaround. The IPC structure is:
COMP_COMMON_DATA
COMP_SPECIFIC_DATA

We now have to add UUID to COMMON_DATA, but unfortunately we only have 4bytes of reserved bits in COMP_COMMON_DATA (sof_ipc_comp). So what @keyonjie has done is:

COMP_COMMON_DATA
COMP_SPECIFIC_DATA
COMP_COMMON_DATA2

... and we use the 32bit reserved field of COMP_COMMON_DATA to refer to start of COMP_COMMON_DATA2.  The new COMP_COMMON_DATA2 is of fixed size, so it's not a flex array. COMP_SPECIFIC_DATA is of variable size, so we cannot just add a flex array member to COMP_COMMON_DATA. This is not really how the IPC protocol was designed to be extended, but to support the ABI, we need to adjust. I couldn't come up with a cleaner way. @marc-hb you proposed to change how the new data at the end is taken into consideration in the header size field, but I think that doesn't change the fundamental layout of the data.

Reply:
@kv2019i @plbossart this was my idea here to bridge us to IPC2.0. The usage of reserved -> offset in COMP_COMMON_DATA to point to COMP_COMMON_DATA2 does allow variable size in COMP_SPECIFIC_DATA.
It also allows for further COMMON data in COMP_COMMON_DATA2 (using reserved fields). This is the best we can do under the circumstances today without a MAJOR ABI update (which will be for IPC2).

Reply:
@lgirdwood I am not pushing back in the idea but on the implementation. The 'offset' is really a size, and I don't understand why we can't use flexible arrays.

Reply:
> The IPC structure is:

Thanks, this short, clear and so far best explanation should really be a in design document, not just buried in a review comment of one of the PRs. I figured it out eventually but only after a couple of hours of very inefficient code review. Of course it was much faster and easier for people familiar with all this we don't want experts only to understand how the IPC protocol is structured at the top(!)-level , do we? 

It's great to see working code but protocols are never defined in just code, that's too tedious and error-prone, they're defined with diagrams like these: https://www.google.com/search?q=dns+header&tbm=isch
This sort of diagrams is _especially_ required when designing backward and forward compatibility which are very tricky.

The reason @kv2019i's explanation is the best I've seen so far is because it's almost a diagram (of what I eventually drew in my head after a too long time)

Reply:
>  I couldn't come up with a cleaner way. @marc-hb you proposed to change how the new data at the end is taken into consideration in the header size field, but I think that doesn't change the fundamental layout of the data.

Correct - which is why I don't think flexible arrays can help either but I'd love to be proven wrong,

I couldn't come up with a cleaner way either but that's only because I'm new to this code and particular protocol, not because there isn't ;-)

Every consistency check that compares some combination of lengths is suspicious because it often points at (unintended) redundancy and I think the `header mismatch!` error above demonstrates just that. Hence my suggested fix. As @plbossart confirmed (thx), the "offset" is really just the total `sizeof(COMP_SPECIFIC_FOO)` so please call it like that and even simpler put it in `comp.hdr.size` which already does this. Then only one bit in the reserved field/flags is enough to indicate the presence of extended data, other bits can be kept and reserved for later.

Reply:
@plbossart I've used "offset" for simplicity. It's really siimple for both FW and kernel. 

The IPC common FW layer that deals with the common data is different from the FW layer/files that deal with the payloads. Only COMP_COMMON_DATA is fixed size, COMP_COMMON_DATA2 and COMP_SPECIFIC data can grow and include var arrays making it more difficuilt for the generic IPC FW layer to deduce where COMP_COMMOM_DATA2 starts without having to import knowledge from further down the stack (about the payload data). i.e. hdr.size is size of everything but COMMON_DATA2 and SPECIFIC_DATA sizes can change.

I want to keep the common IPC and comp IPC separated in FW as this boundary also simplifies effort for IPC2. I know these layers are converged in the kernel (topology.c), and we should be able to really simplify them with IPC2 (we need a new topology2.c which only deals with 7 IPC structures in total - more to come on this later).

Reply:
I don't mind using offset and deducting the size of the additional payload as header.size-offset, but that's not how I understood the driver code. I'd really like a PR where definitions and code don't seem to contradict each other or mislead reviewers. Is this too much to ask?

Reply:
@plbossart agree, this should be obvious from the driver too.
@keyonjie can you make this really obvious by adding new doxygen comments in the IPC header alongside appropriate logic/comments in the driver. Thanks.

Reply:
> .e. hdr.size is size of everything but COMMON_DATA2 and SPECIFIC_DATA sizes can change.

I believe the sizes are now:
```
- comp.hdr.size          =  COMP_COMMON + comp_specific + common_data2
- extended_data "offset" =  COMP_COMMON + comp_specific
```
I'd like to understand what's wrong about the simpler:
```
- comp.hdr.size       = COMP_COMMON + comp_specific
- extended_data SIZE  = common_data2 
```

Reply:
> > .e. hdr.size is size of everything but COMMON_DATA2 and SPECIFIC_DATA sizes can change.
> 
> I believe the sizes are now:
> 
> ```
> - comp.hdr.size          =  COMP_COMMON + comp_specific + common_data2

here the size of common_data2 is from the Linux driver side. We can read the correct 'source' size with it via:
```
	/* read rest of component data */
	mailbox_hostbox_read(hdr + 1, SOF_IPC_MSG_MAX_SIZE - sizeof(*hdr),
			     sizeof(*hdr), hdr->size - sizeof(*hdr));
```

> - extended_data "offset" =  COMP_COMMON + comp_specific
> ```
> 
> I'd like to understand what's wrong about the simpler:
> 
> ```
> - comp.hdr.size       = COMP_COMMON + comp_specific
> - extended_data SIZE  = common_data2 

Here the common_data2 is defined in the Firmware, this could be different with the one from the Linux driver side if the ABI is not exactly matched, e.g. if the FW is newer than the driver, it will read for more bytes, which means you will read bytes not from source at this case...
```
	/* read rest of component data */
	mailbox_hostbox_read(hdr + 1, SOF_IPC_MSG_MAX_SIZE - sizeof(*hdr),
			     sizeof(*hdr), hdr->size - sizeof(*hdr) + sizeof(struct sof_ipc_comp_new_ext)); // Keyon: the sizeof(struct sof_ipc_comp_new_ext) could be not accurate when ABI mismatched.
```
> ```

Reply:
> @plbossart agree, this should be obvious from the driver too.
> @keyonjie can you make this really obvious by adding new doxygen comments in the IPC header alongside appropriate logic/comments in the driver. Thanks.

Hi @plbossart @lgirdwood there is already some related comments in the IPC header definitions, I can add something like below if you prefer:

```
/**
 * Command Header - Header for all IPC commands. Identifies IPC message.
 * The size can be greater than the structure size and that means there is
 * extended bespoke data beyond the end of the structure including variable
 * arrays.
 */
 struct sof_ipc_cmd_hdr {
-       uint32_t size;                  /**< size of structure */
+       /**< size of whole IPC structure, including bespoke data */
+       uint32_t size;
        uint32_t cmd;                   /**< SOF_IPC_GLB_ + cmd */
 } __attribute__((packed));
```

@plbossart which code is contradict to the definitions above did you refer to?
In the driver side, we are using the naming 'ipc_size' in each sof_widget_load_xxx(), if you want, I can add also similar comments to each of it, e.g.:
```
+      /* size of whole IPC structure, including any bespoke data and extended data */
	size_t ipc_size = sizeof(*mux) + sizeof(swidget->comp_ext);
```

Reply:
> this could be different with the one from the Linux driver side if the ABI is not exactly matched, e.g. if the FW is newer than the driver,

Yes.

> it will read for more bytes,

No:

```C
/* read rest of component data */
mailbox_hostbox_read(hdr + 1, SOF_IPC_MSG_MAX_SIZE - sizeof(*hdr),
		    sizeof(*hdr), hdr->size - sizeof(*hdr) + comp.ext_data_SIZE)
```

You're mixing low-level implementation details with high-level protocol design. That's because the high-level protocol design and especially the tricky forward/backward compatibility have not been documented and reviewed outside the code and keep being inferred from it.

> * The size can be greater than the structure size and that means there is
>  * extended bespoke data beyond the end of the structure including variable
>  * arrays.

AFAIK there's nothing the C compiler sees as variable. "variable arrays" sounds too close to "VLA", this should be rephrased.

Reply:
@keyonjie please complet the below implementation - see the changes to the calculations.
```c
static const struct comp_driver *get_drv(struct sof_ipc_comp *comp)
{
	struct comp_driver_list *drivers = comp_drivers_get();
	struct list_item *clist;
	const struct comp_driver *drv = NULL;
	struct comp_driver_info *info;
	struct sof_ipc_comp_new_ext *comp_ext;
	uint32_t comp_ext_size;
	uint32_t flags;

	irq_local_disable(flags);

	/* do we have extended data ? */
	if (!comp->ext_data_offset)
		goto comp_type_match;
	
	
	/* we have extended data  - validate the data offset */
	if (comp->ext_data_offset >= comp->hdr.size -
	    offsetof(struct sof_ipc_comp, ext_data_offset) {
		tr_err(&comp_tr, "get_drv(): offset points beyond structure: ext_data_offset:0x%x, hdr.size: 0x%x",
		       comp->ext_data_offset, comp->hdr.size);
		return NULL;
	}

	/* extended data position is relative to the offset member i.e. "from here" */ 
	comp_ext = (struct sof_ipc_comp_new_ext *)
		   ((uint8_t *)(comp->ext_data_offset) + comp->ext_data_offset);

	/* size of extended data */
	comp_ext_size = comp->hdr.size + offsetof(struct sof_ipc_comp, ext_data_offset) -
			comp->ext_data_offset;

	/* UUID is first item in extended data - check its big enough */
	if (comp_ext_size < uuid_size) {
		tr_warn(&comp_tr, "UUID is invalid, please check your tplg file, component type will be used\n");
		goto comp_type_match;
	}

	/* search driver list with UUID */
	list_for_item(clist, &drivers->list) {
		info = container_of(clist, struct comp_driver_info,
				    list);
		if (!memcmp(info->drv->uid, comp_ext->uuid,
			    UUID_SIZE)) {
			tr_dbg(&comp_tr,
			       "get_drv_from_uuid(), found driver type %d, uuid %s",
			       info->drv->type,
			       info->drv->tctx->uuid_p);
			drv = info->drv;
			return drv;
		}
	}

	tr_err(&comp_tr, "get_drv(): the provided UUID (%8x%8x%8x%8x) doesn't match to any driver!",
	       *(uint32_t *)(&comp_ext->uuid[0]),
	       *(uint32_t *)(&comp_ext->uuid[4]),
	       *(uint32_t *)(&comp_ext->uuid[8]),
	       *(uint32_t *)(&comp_ext->uuid[12]));
	       
	       
	/* new extended data parsing can be added here by checking the remaining size
	 * and order each item is added by ABI */

	return NULL;
	

comp_type_match:
	/* search driver list for driver type */
	list_for_item(clist, &drivers->list) {
		info = container_of(clist, struct comp_driver_info, list);
		if (info->drv->type == comp->type) {
			drv = info->drv;
			platform_shared_commit(info, sizeof(*info));
			goto out;
		}

		platform_shared_commit(info, sizeof(*info));
	}

out:
	platform_shared_commit(drivers, sizeof(*drivers));
	irq_local_enable(flags);
	return drv;
}
```

Reply:
All: I assume we now go with the above definition. I'll add it to the the IPC doc PR as well.

Reply:
@lgirdwood @keyonjie I think there's one bug in Liam's pseudocode:
-	comp_ext_size = comp->hdr.size + offsetof(struct sof_ipc_comp, ext_data_offset) -
-			comp->ext_data_offset;
+	comp_ext_size = comp->hdr.size - offsetof(struct sof_ipc_comp, ext_data_offset) -
+			comp->ext_data_offset;

Otherwise the start of IPC COMP message is calculated to comp_ext_size.

Reply:
Assuming I didn't get lost and understood correctly the most recent change, now @lgirdwood you basically want the `reserved` field to hold the size added by comp_specific fields plus/minus 4 bytes, correct? Then just make it and call it something like `specific_fields_length` because sizes and lengths are much easier to understand than offsets. Lengths and sizes are easier because a size has only one parameter whereas an offset is always between _two_ different points, so it's higher effort thinking and more error prone.

> I think there's one bug in Liam's pseudocode:

There. Phase 2 of #3143 can't come too soon.

In other words:

```C
if (extended) // extended data is located immediately after specific fields
  specific_fields_length = sizeof(derived comp class) - sizeof(base comp class)
else // backward compatibility with code before extended data
  specific_fields_length = 0
```



Please also rename the now totally vague `comp->hdr.size` to a more specific and meaningful `comp->hdr.something_size`

Reply:
> @lgirdwood @keyonjie I think there's one bug in Liam's pseudocode:
> 
> ```
> -	comp_ext_size = comp->hdr.size + offsetof(struct sof_ipc_comp, ext_data_offset) -
> -			comp->ext_data_offset;
> +	comp_ext_size = comp->hdr.size - offsetof(struct sof_ipc_comp, ext_data_offset) -
> +			comp->ext_data_offset;
> ```
> 
> Otherwise the start of IPC COMP message is calculated to comp_ext_size.

Thank you @kv2019i  that is correct.

@marc-hb I dont want large impact name changes in IPC1.x here, lets keep the header struct member naming as is today, but we can add more inline comments. We can make changes to make names more obvious in IPC2 though.
I'm fine with changes to offset -> extended length providing @keyonjie fully understands this ?

Reply:
> I dont want large impact name changes in IPC1.x here, lets keep the header struct member naming as is today, but we can add more inline comments.

Sorry I didn't expect `hdr.size` to be used in so many places. More doxygen works.

> I'm fine with changes to offset -> extended length providing @keyonjie fully understands this ?

After a bit more time I think `derived_members_size` would be a clear name because it's the exact object oriented terminology. "Members" is the name used by doxygen: https://thesofproject.github.io/latest/api/component-api.html

In case that's too much OO language then `specific_members_size` would be clear too.

Reply:
Thanks @marc-hb for suggestion, it looks good to me except that the naming 'extended (or specific field) length' only means length, even when there is no extended data, the specific field is still existed and we should not set this length to be 0, so we can't use this length to denote if there is extended data appended or not?